### ISSUE TITLE
[Linux] SDK fails to start on RHEL 7.9 since I20241127-1240, requires GLIBC_2.34

### ISSUE THREAD

**Issue by iloveeclipse** [eclipse-platform/eclipse.platform.swt#1631]
Either I've missed the announcement or we have an issue with supporting older Linux installations with 4.35.
I'm unable to start latest IBuild on RHEL 7.9, same build works fine on RHEL 9.2.

IDE shows splash screen and after that an error dialog.
![image](https://github.com/user-attachments/assets/a114f956-f9e7-48cf-a10a-27cbfc804395)

Log shows that we have an implicit requirement of `GLIBC_2.34` now.
I have `GLIBC_2.17` (`glibc-2.17-317.el7.i686`) on RHEL 7.9:
```
rpm -qa | sort | grep glibc
glibc-2.17-317.el7.i686
glibc-2.17-317.el7.x86_64
glibc-common-2.17-317.el7.x86_64
glibc-debuginfo-2.17-317.el7.x86_64
glibc-debuginfo-common-2.17-317.el7.x86_64
glibc-devel-2.17-317.el7.i686
glibc-devel-2.17-317.el7.x86_64
glibc-headers-2.17-317.el7.x86_64
```

```
!SESSION 2024-12-02 13:50:14.865 -----------------------------------------------
eclipse.buildId=4.35.0.I20241201-1800
java.version=21.0.2-13
java.vendor=N/A
BootLoader constants: OS=linux, ARCH=x86_64, WS=gtk, NL=en_US
Command-line arguments:  -os linux -ws gtk -arch x86_64

!ENTRY org.eclipse.osgi 4 0 2024-12-02 13:50:30.814
!MESSAGE Application error
!STACK 1
java.lang.UnsatisfiedLinkError: Could not load SWT library. Reasons: 
	/home/aloskuto/Downloads/eclipse435/eclipse/configuration/org.eclipse.osgi/267/0/.cp/libswt-pi3-gtk-4968r2.so: /lib64/libc.so.6: version `GLIBC_2.34' not found (required by /home/aloskuto/Downloads/eclipse435/eclipse/configuration/org.eclipse.osgi/267/0/.cp/libswt-pi3-gtk-4968r2.so)
	no swt-pi3-gtk in java.library.path: /opt/rh/devtoolset-11/root/usr/lib64:/opt/rh/devtoolset-11/root/usr/lib:/opt/rh/devtoolset-11/root/usr/lib64/dyninst:/workspaces/socbm262/aloskuto/ws-r7.10-aloskuto-bf/vobs/zenith/workspace/lib:/workspaces/socbm262/aloskuto/ws-r7.10-aloskuto-bf/opt/93000/src/fw/hpl/sh_lib:/workspaces/socbm262/aloskuto/ws-r7.10-aloskuto-bf/opt/93000/src/segments/eclipse/eclipseRE/sh_lib:/opt/hp93000rt/el7/x86_64/ure_udk_74/sdk/lib:/opt/hp93000rt/el7/x86_64/ure_udk_74/ure/lib:/opt/hp93000rt/el7/x86_64/ace-tao_6.2.4/ACE/lib64:/opt/hp93000rt/el7/x86_64/ruby-1.8.6/lib:/opt/hp93000rt/el7/x86_64/log4cxx_0.11.0_snapshot/lib:/opt/hp93000rt/el7/x86_64/imkl_3:/opt/hp93000rt/el7/x86_64/libstdf_0.4/lib:/usr/java/packages/lib:/usr/lib64:/lib64:/lib:/usr/lib
	no swt-pi3 in java.library.path: /opt/rh/devtoolset-11/root/usr/lib64:/opt/rh/devtoolset-11/root/usr/lib:/opt/rh/devtoolset-11/root/usr/lib64/dyninst:/workspaces/socbm262/aloskuto/ws-r7.10-aloskuto-bf/vobs/zenith/workspace/lib:/workspaces/socbm262/aloskuto/ws-r7.10-aloskuto-bf/opt/93000/src/fw/hpl/sh_lib:/workspaces/socbm262/aloskuto/ws-r7.10-aloskuto-bf/opt/93000/src/segments/eclipse/eclipseRE/sh_lib:/opt/hp93000rt/el7/x86_64/ure_udk_74/sdk/lib:/opt/hp93000rt/el7/x86_64/ure_udk_74/ure/lib:/opt/hp93000rt/el7/x86_64/ace-tao_6.2.4/ACE/lib64:/opt/hp93000rt/el7/x86_64/ruby-1.8.6/lib:/opt/hp93000rt/el7/x86_64/log4cxx_0.11.0_snapshot/lib:/opt/hp93000rt/el7/x86_64/imkl_3:/opt/hp93000rt/el7/x86_64/libstdf_0.4/lib:/usr/java/packages/lib:/usr/lib64:/lib64:/lib:/usr/lib
	Can't load library: /home/aloskuto/.swt/lib/linux/x86_64/libswt-pi3-gtk-4968r2.so
	Can't load library: /home/aloskuto/.swt/lib/linux/x86_64/libswt-pi3-gtk.so
	Can't load library: /home/aloskuto/.swt/lib/linux/x86_64/libswt-pi3.so
	/home/aloskuto/.swt/lib/linux/x86_64/libswt-pi3-gtk-4968r2.so: /lib64/libc.so.6: version `GLIBC_2.34' not found (required by /home/aloskuto/.swt/lib/linux/x86_64/libswt-pi3-gtk-4968r2.so)

	at org.eclipse.swt.internal.Library.loadLibrary(Library.java:345)
	at org.eclipse.swt.internal.Library.loadLibrary(Library.java:254)
	at org.eclipse.swt.internal.gtk.OS.<clinit>(OS.java:99)
	at org.eclipse.swt.internal.Converter.wcsToMbcs(Converter.java:209)
	at org.eclipse.swt.internal.Converter.wcsToMbcs(Converter.java:155)
	at org.eclipse.swt.widgets.Display.<clinit>(Display.java:170)
	at org.eclipse.ui.internal.Workbench.createDisplay(Workbench.java:754)
	at org.eclipse.ui.PlatformUI.createDisplay(PlatformUI.java:185)
	at org.eclipse.ui.internal.ide.application.IDEApplication.createDisplay(IDEApplication.java:208)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:151)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:208)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:143)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:109)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:439)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:271)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:668)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:605)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1481)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1454)
	Suppressed: java.lang.UnsatisfiedLinkError: Could not load SWT library. Reasons: 
	/home/aloskuto/Downloads/eclipse435/eclipse/configuration/org.eclipse.osgi/267/0/.cp/libswt-pi4-gtk-4968r2.so: libgtk-4.so.1: cannot open shared object file: No such file or directory
	no swt-pi4-gtk in java.library.path: /opt/rh/devtoolset-11/root/usr/lib64:/opt/rh/devtoolset-11/root/usr/lib:/opt/rh/devtoolset-11/root/usr/lib64/dyninst:/workspaces/socbm262/aloskuto/ws-r7.10-aloskuto-bf/vobs/zenith/workspace/lib:/workspaces/socbm262/aloskuto/ws-r7.10-aloskuto-bf/opt/93000/src/fw/hpl/sh_lib:/workspaces/socbm262/aloskuto/ws-r7.10-aloskuto-bf/opt/93000/src/segments/eclipse/eclipseRE/sh_lib:/opt/hp93000rt/el7/x86_64/ure_udk_74/sdk/lib:/opt/hp93000rt/el7/x86_64/ure_udk_74/ure/lib:/opt/hp93000rt/el7/x86_64/ace-tao_6.2.4/ACE/lib64:/opt/hp93000rt/el7/x86_64/ruby-1.8.6/lib:/opt/hp93000rt/el7/x86_64/log4cxx_0.11.0_snapshot/lib:/opt/hp93000rt/el7/x86_64/imkl_3:/opt/hp93000rt/el7/x86_64/libstdf_0.4/lib:/usr/java/packages/lib:/usr/lib64:/lib64:/lib:/usr/lib
	no swt-pi4 in java.library.path: /opt/rh/devtoolset-11/root/usr/lib64:/opt/rh/devtoolset-11/root/usr/lib:/opt/rh/devtoolset-11/root/usr/lib64/dyninst:/workspaces/socbm262/aloskuto/ws-r7.10-aloskuto-bf/vobs/zenith/workspace/lib:/workspaces/socbm262/aloskuto/ws-r7.10-aloskuto-bf/opt/93000/src/fw/hpl/sh_lib:/workspaces/socbm262/aloskuto/ws-r7.10-aloskuto-bf/opt/93000/src/segments/eclipse/eclipseRE/sh_lib:/opt/hp93000rt/el7/x86_64/ure_udk_74/sdk/lib:/opt/hp93000rt/el7/x86_64/ure_udk_74/ure/lib:/opt/hp93000rt/el7/x86_64/ace-tao_6.2.4/ACE/lib64:/opt/hp93000rt/el7/x86_64/ruby-1.8.6/lib:/opt/hp93000rt/el7/x86_64/log4cxx_0.11.0_snapshot/lib:/opt/hp93000rt/el7/x86_64/imkl_3:/opt/hp93000rt/el7/x86_64/libstdf_0.4/lib:/usr/java/packages/lib:/usr/lib64:/lib64:/lib:/usr/lib
	Can't load library: /home/aloskuto/.swt/lib/linux/x86_64/libswt-pi4-gtk-4968r2.so
	Can't load library: /home/aloskuto/.swt/lib/linux/x86_64/libswt-pi4-gtk.so
	Can't load library: /home/aloskuto/.swt/lib/linux/x86_64/libswt-pi4.so
	/home/aloskuto/.swt/lib/linux/x86_64/libswt-pi4-gtk-4968r2.so: libgtk-4.so.1: cannot open shared object file: No such file or directory

		at org.eclipse.swt.internal.Library.loadLibrary(Library.java:345)
		at org.eclipse.swt.internal.Library.loadLibrary(Library.java:254)
		at org.eclipse.swt.internal.gtk.OS.<clinit>(OS.java:103)
		... 18 more
```

I assume this is regression from either ca1bb19913c96121ef58a79a3f2660332d5b98fd or c984d148cf1471cb9dd31a76e60d6a8ada7594c7

**Comment by iloveeclipse** [comment]
@akurtakov, @laeubi : at Advantest we are forced to provide support for older Linux versions because our product runs (once adopted by customers) typically for many years in **unchanged** production environment. So we are supporting RHEL 7.9 till end of 2025 at least, even if we **also** support **same** product version running on RHEL 9.2.

Dropping RHEL 7.9 support would be not nice, at least till 2026.

I **assume** that the glibc requirement is coming from compilation on latest ubuntu (https://github.com/eclipse-platform/eclipse.platform.swt/pull/1624), as I don't see **explicit** requirement on GLIBC 2.34 in the SWT sources.

Isn't there a way to compile GTK3 related SWT code in a way it doesn't require GLIBC 2.34?

**Comment by akurtakov** [comment]
Compilation actually happens on debian 12  for https://github.com/eclipse-platform/eclipse.platform.swt/pull/1422 . Waiting for 2026 to start providing gtk 4 builds looks like the final nail for swt so I don't have a clue how I can help you.

**Comment by iloveeclipse** [comment]
@laeubi : thanks. I see your message says:

> As part of this effort it was required to update some build dependencies resulting in a higher version of GLIBC (at least 2.36 for x86 other architectures might require higher versions) this usually means you need at least:
> - Debian 12 or recent Ubuntu
> - Centos 9

On RHEL 9.2 (from which CentOS 9 is built) we have GLIBC **2.34** which is below of GLIBC **2.36**  what you've mentioned.

Is this expected that RHEL 9.2 is also not supported and works "by chance" now?

**Comment by laeubi** [comment]
> On RHEL 9.2 (from which CentOS 9 is built) we have GLIBC 2.34 which is below of GLIBC 2.36 what you've mentioned.
> Is this expected that RHEL 9.2 is also not supported and works "by chance" now?

The main problem is that CentOS 9 is no longer a "stable" release but a roiling one (from what I have understood), the closes we could found compared to what we previously had with centos was Debian LTS releases and that's what we compile against, and running against latest ubuntu, so there is absolutely possible that RHEL/Centos only "work by chance" unless there is some efforts to even run the tests on different distributions.

At least you have verified now it runs on RHEL 9.2 and @akurtakov uses Fedora, not sure if we have more variations (I test under Debian 12 directly).

**Comment by laeubi** [comment]
> This is new, not sure your are on the right track. We have chosen RHEL because of the enterprise support and platform stability guaranteed. Rolling release is per definition not stable so from my experience since (RHEL 5) I doubt RHEL 9 is "rolling" release.

RHEL 9 != Centos 9 but I'm not that deep in that topic and its mostley marketing / bussineess descision I'm not involved in.

> But once again: is GLIBC 2.36 the minimum required version now? If yes, RHEL 9.2 is not supported then?

according to the error you send here and with this command (thanks to @akurtakov for the hint) :

```
/git/eclipse.platform.swt/binaries/org.eclipse.swt.gtk.linux.x86_64$ nm -D *.so |grep GLIBC|sort -u
                 U dlopen@GLIBC_2.34
                 U dlsym@GLIBC_2.34
                 U fflush@GLIBC_2.2.5
                 U fprintf@GLIBC_2.2.5
                 U free@GLIBC_2.2.5
                 U getenv@GLIBC_2.2.5
                 U localeconv@GLIBC_2.2.5
                 U malloc@GLIBC_2.2.5
                 U memcpy@GLIBC_2.14
                 U memmove@GLIBC_2.2.5
                 U memset@GLIBC_2.2.5
                 U realpath@GLIBC_2.3
                 U setenv@GLIBC_2.2.5
                 U stderr@GLIBC_2.2.5
                 U strcmp@GLIBC_2.2.5
                 U strlen@GLIBC_2.2.5
                 U strtok@GLIBC_2.2.5
                 w __cxa_finalize@GLIBC_2.2.5
```

actually **2.34** is the minimum (as of today!)

**Comment by akurtakov** [comment]
@iloveeclipse Times have changed ( https://blog.centos.org/2020/12/future-is-centos-stream/ ) and RHEL is no longer upstream for CentOS but rather the opposite.

**Comment by iloveeclipse** [comment]
For RHEL 7 we will evaluate a (local) rebuild of SWT natives on RHEL 7.

For the record, RH customer portal lists following versions of glibc for **RHEL 9.4* (latest released):
![image](https://github.com/user-attachments/assets/4599a2e8-9def-4e32-a17b-d207ec0cfd34)
The **only available choice** is GLIBC **2.34**.

So it would be nice we could **at least** keep that level in SWT.

**Comment by akurtakov** [comment]
Closing as this is mandatory step to keep SWT alive. Documentation enhancements are to be handled in the referenced issue.
